engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/DelayedKafkaAvroSourceFactory.scala

```diff
@@ -0,0 +1,82 @@
+                                  (implicit nodeId: ProcessCompilationError.NodeId): NodeTransformationDefinition = {
+    case step@TransformationStep(
+      (`topicParamName`, DefinedEagerParameter(topic: String, _)) ::
+        (SchemaVersionParamName, DefinedEagerParameter(version: String, _)) ::
+        (TimestampFieldParamName, DefinedEagerParameter(field, _)) ::
+        (DelayParameterName, DefinedEagerParameter(delay, _)) :: Nil, _) =>
+      val preparedTopic = prepareTopic(topic)
+      val versionOption = parseVersionOption(version)
+      val valueValidationResult = determineSchemaAndType(prepareValueSchemaDeterminer(preparedTopic, versionOption), Some(SchemaVersionParamName))
+
+      valueValidationResult match {
+        case Valid((valueRuntimeSchema, _)) =>
+          val definition = valueRuntimeSchema.map(_.schema.getFields.asScala.map(f => (f.name(), ())).toMap ).getOrElse(Map.empty).asJava
+          val delayValidationErrors = Option(delay.asInstanceOf[java.lang.Long]).map(d => validateDelay(d)).getOrElse(Nil)
+          val timestampValidationErrors = Option(field.asInstanceOf[String]).map(f => validateTimestampField(f, definition)).getOrElse(Nil)
+          val errors = delayValidationErrors ++ timestampValidationErrors
+          prepareSourceFinalResults(preparedTopic, valueValidationResult, context, dependencies, step.parameters, errors)
+        case Invalid(exc) =>
+          prepareSourceFinalErrors(context, dependencies, step.parameters, List(exc))
+      }
+    case step@TransformationStep((`topicParamName`, _) :: (SchemaVersionParamName, _) :: (TypeDefinitionParamName, _) :: (TimestampFieldParamName, _) :: (DelayParameterName, _) :: Nil, _) =>
+      prepareSourceFinalErrors(context, dependencies, step.parameters, errors = Nil)
+  }
+
+  override protected def createSource(params: Map[String, Any],
+                                      dependencies: List[NodeDependencyValue],
+                                      finalState: Option[State],
+                                      preparedTopics: List[PreparedKafkaTopic],
+                                      kafkaConfig: KafkaConfig,
+                                      deserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]],
+                                      timestampAssigner: Option[TimestampWatermarkHandler[ConsumerRecord[K, V]]],
+                                      formatter: RecordFormatter,
+                                      flinkContextInitializer: FlinkContextInitializer[ConsumerRecord[K, V]]): KafkaSource[ConsumerRecord[K, V]] = {
+    val timestampFieldName = extractTimestampField(params)
+    val extractTimestamp: (ConsumerRecord[K, V], Long) => Long = (consumerRecord, kafkaEventTimestamp) => {
+      Option(timestampFieldName)
+        .map(f => consumerRecord.value().asInstanceOf[GenericRecord].get(f).asInstanceOf[Long])
+        .getOrElse(kafkaEventTimestamp)
+    }
```

Comment:
1. We should keep logic of timestamp extraction consistent - if user define timestamp extraction strategy, it should override timestamp assignment logic. The best would be passing timestamp assigner to `DelayedFlinkKafkaConsumer`
2. Why not to use LazyParameter for timestamp instead of fixing this implementation to `GenericRecord`? 
3. Let's add TODO that this timestamp extraction logic configured in parameter (instead of events normaliation in ETLs + configured timestamp assigner) should be also reused in normal kafka (avro/json) sources